### PR TITLE
[lld-macho] Fix compatibility between --icf=safe_thunks and --keep-icf-stabs

### PR DESCRIPTION
**Pull request by alx32**
Currently when `--icf=safe_thunks` is used, `STABS` entries cannot be generated for ICF'ed functions. This is because if ICF converts a full function into a thunk and then we generate a `STABS` entry for the thunk, `dsymutil` will expect to find the entire function body at the location of the thunk. Because just a thunk will be present at the location of the `STABS`  entry - dsymutil will generate invalid debug info for such scenarios.

With this change, if `--icf=safe_thunks` is used and `--keep-icf-stabs` is also specified, STABS entries will be created for all functions, even merged ones. However, the STABS entries will point at the actual (full) function body while having the name of the thunk. This way we still get program correctness as well as correct DWARF data. When doing this, the debug data will be identical to the scenario where we're using `--icf=all` and `--keep-icf-stabs`, but the actual program will also contain thunks, which won't show up in the DWARF data.

### CODE REVIEW

File: lld/MachO/ICF.cpp [hunk 1]
```diff
@@ -481,6 +481,34 @@ void macho::markAddrSigSymbols() {
...
+
+  for (Symbol *sym : functionBody->symbols) {
+    Defined *d = dyn_cast<Defined>(sym);
+    // The symbol needs to be at the start of the InputSection
+    if (d && d->value == 0)
+      return d;
+  }
+
+  llvm_unreachable("could not find body symbol for ICF-generated thunk");
+  return nullptr;
```
Comment (ellishg) [inline lld/MachO/ICF.cpp:510]: This is not reachable.
```suggestion
```

File: lld/MachO/SyntheticSections.cpp [hunk 2]
```diff
@@ -1251,26 +1255,26 @@ void SymtabSection::emitStabs() {
 
       // We use 'originalIsec' to get the file id of the symbol since 'isec()'
       // might point to the merged ICF symbol's file
-      symbolsNeedingStabs.emplace_back(defined,
-                                       defined->originalIsec->getFile()->id);
+      Defined *funcBodySym = getFuncBodySym(defined);
+      symbolsNeedingStabs.emplace_back(SymbolStabInfo{
+          defined, funcBodySym->originalIsec->getFile()->id, funcBodySym});
```
Comment (ellishg) [inline lld/MachO/SyntheticSections.cpp:1260]: I could be wrong, but I think this isn't needed.
```suggestion
      symbolsNeedingStabs.emplace_back(
          defined, funcBodySym->originalIsec->getFile()->id, funcBodySym);
```

File: lld/MachO/SyntheticSections.cpp [hunk 3]
```diff
@@ -1251,26 +1255,26 @@ void SymtabSection::emitStabs() {
 
       // We use 'originalIsec' to get the file id of the symbol since 'isec()'
       // might point to the merged ICF symbol's file
-      symbolsNeedingStabs.emplace_back(defined,
-                                       defined->originalIsec->getFile()->id);
+      Defined *funcBodySym = getFuncBodySym(defined);
```
Comment (ellishg) [inline lld/MachO/SyntheticSections.cpp:1258]: Do you plan to use `getFuncBodySym()` elsewhere? Because I think it'll be more readable to inline it here.
```suggestion
      Defined *funcBodySym = defined->identicalCodeFoldingKind == Symbol::ICFFoldKind::Thunk ? macho::getBodyForThunkFoldedSym(defined) : defined;
```

File: lld/MachO/SyntheticSections.cpp [hunk 4]
```diff
@@ -1251,26 +1255,26 @@ void SymtabSection::emitStabs() {
...
+                      return a.fileId < b.fileId;
                     });
 
   // Emit STABS symbols so that dsymutil and/or the debugger can map address
   // regions in the final binary to the source and object files from which they
   // originated.
   InputFile *lastFile = nullptr;
-  for (SortingPair &pair : symbolsNeedingStabs) {
-    Defined *defined = pair.first;
+  for (const SymbolStabInfo &info : symbolsNeedingStabs) {
```
Comment (ellishg) [inline lld/MachO/SyntheticSections.cpp:1273]: Why do we need to maintain `mainBodySym` in `SymbolStabInfo`? Could we just compute it here?

```suggestion
  for (const SymbolStabInfo &info : symbolsNeedingStabs) {
    Defined *funcBodySym = originalSym->identicalCodeFoldingKind == Symbol::ICFFoldKind::Thunk ? macho::getBodyForThunkFoldedSym(originalSym) : originalSym;
```

File: lld/MachO/ICF.cpp (hunk 1 above)
Comment (DataCorrupted) [inline lld/MachO/ICF.cpp:510]: Is return required after an `llvm_unreachable()`?

**Review by ellishg — COMMENTED** [review]
Also CC @thevinster and @keith who have worked in these files before.

File: lld/MachO/ICF.cpp (hunk 1 above)
Comment (alx32) [reply lld/MachO/ICF.cpp:510]: I see both [yes](https://github.com/llvm/llvm-project/blob/main/bolt/include/bolt/Core/MCPlusBuilder.h#L419) and [no](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/index/FileIndex.cpp#L400) in the codebase. I think technically it might not be required, but I was thinking this is the 'safe' way to go.

File: lld/MachO/SyntheticSections.cpp (hunk 4 above)
Comment (alx32) [reply lld/MachO/SyntheticSections.cpp:1273]: We need `funcBodySym->originalIsec->getFile()->id` in `SymbolStabInfo` in order to sort. Since we already have `funcBodySym` above, I decided to store it in `SymbolStabInfo` to not re-compute it again later.

File: lld/MachO/SyntheticSections.cpp (hunk 2 above)
Comment (alx32) [reply lld/MachO/SyntheticSections.cpp:1260]: I think that works if `SymbolStabInfo` has a 3-param constructor. As-is it'll give errors like:
```
candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 3 were provided
```

File: lld/MachO/ICF.cpp (hunk 1 above)
Comment (alx32) [reply lld/MachO/ICF.cpp:510]: Commented above also:

I see both [yes](https://github.com/llvm/llvm-project/blob/main/bolt/include/bolt/Core/MCPlusBuilder.h#L419) and [no](https://github.com/llvm/llvm-project/blob/main/clang-tools-extra/clangd/index/FileIndex.cpp#L400) in the codebase. I think technically it might not be required, but I was thinking this is the 'safe' way to go.

Any preference on how to proceed ?

File: lld/MachO/SyntheticSections.cpp (hunk 3 above)
Comment (ellishg) [reply lld/MachO/SyntheticSections.cpp:1258]: The logic seems pretty clear IMO: "If this is a thunk, call this special thunk function".

If you want to keep the function, then maybe we should rename it and add some docs. In the most common case, when `defined` is not a thunk, it is very mysterious why we need to pass it through `getFuncBodySym()` first.

File: lld/MachO/SyntheticSections.cpp (hunk 4 above)
Comment (ellishg) [reply lld/MachO/SyntheticSections.cpp:1273]: And the cache would only have entries for thunks so we get a space savings.

File: lld/MachO/SyntheticSections.cpp [hunk 5]
```diff
@@ -1214,9 +1226,12 @@ void SymtabSection::emitStabs() {
...
 
-  // Cache the file ID for each symbol in an std::pair for faster sorting.
-  using SortingPair = std::pair<Defined *, int>;
-  std::vector<SortingPair> symbolsNeedingStabs;
+  struct SymbolStabInfo {
+    Defined *originalSym; // Original Defined symbol - this may be an ICF thunk
+    int fileId;           // File ID associated with the STABS symbol
+  };
+
+  std::vector<SymbolStabInfo> symbolsNeedingStabs;
```
Comment (ellishg) [inline lld/MachO/SyntheticSections.cpp:1234]: Now that you haven't added a new field, these changes are unnecessary.

File: lld/MachO/SyntheticSections.cpp [hunk 6]
```diff
@@ -1251,26 +1256,27 @@ void SymtabSection::emitStabs() {
...
   InputFile *lastFile = nullptr;
-  for (SortingPair &pair : symbolsNeedingStabs) {
-    Defined *defined = pair.first;
+  for (const SymbolStabInfo &info : symbolsNeedingStabs) {
     // We use 'originalIsec' of the symbol since we care about the actual origin
     // of the symbol, not the canonical location returned by `isec()`.
-    InputSection *isec = defined->originalIsec;
-    ObjFile *file = cast<ObjFile>(isec->getFile());
+    Defined *funcBodySym = getFuncBodySym(info.originalSym);
+    InputSection *bodyIsec = funcBodySym->originalIsec;
```
Comment (ellishg) [inline lld/MachO/SyntheticSections.cpp:1278]: Why do we need to change the name of `isec`? It seems to create unnecessary changes below.

**Review by ellishg — COMMENTED** [review]
LGTM, but I'll give others more familiar with STABs time to accept

File: lld/test/MachO/icf-safe-thunks-dwarf.ll [hunk 7]
```diff
@@ -20,6 +20,59 @@
 ; VERIFY-STABS:  N_FUN{{.*}}_func_A
 ; VERIFY-STABS:  N_FUN{{.*}}_take_func_addr
 
+;;;;;;;;;;;;;;;;;;;;;;;;;;;; Check safe_thunks ICF + keeping STABS entries ;;;;;;;;;;;;;;;;;;;;;;;;;;;;
+
+;;; Check scenario with where we do safe_thunks ICF and also generate STABS entries
+; RUN: %lld -arch arm64 -lSystem --icf=safe_thunks --keep-icf-stabs -dylib -o %t/a_thunks.dylib %t/a.o
+; RUN: dsymutil -s %t/a_thunks.dylib > %t/a_thunks.txt
```
Comment (kyulee-com) [inline lld/test/MachO/icf-safe-thunks-dwarf.ll:27]: `%t/a_thunks.txt` seems being overridden by another dsymutil call below.

File: lld/MachO/ICF.cpp [hunk 8]
```diff
@@ -481,6 +481,33 @@ void macho::markAddrSigSymbols() {
...
+         "thunk-folded ICF symbol expected to be on a ConcatInputSection");
+  // foldedSec is the InputSection that was marked as deleted upon fold
+  ConcatInputSection *foldedSec =
+      cast<ConcatInputSection>(foldedSym->originalIsec);
+
+  // thunkBody is the actual live thunk, containing the code that branches to
+  // the actual body of the function.
+  InputSection *thunkBody = foldedSec->replacement;
+
+  // the actual (merged) body of the function that the thunk jumps to. This will
```
Comment (kyulee-com) [inline lld/MachO/ICF.cpp:498]:
```suggestion
  // The actual (merged) body of the function that the thunk jumps to. This will
```

File: lld/test/MachO/icf-safe-thunks-dwarf.ll (hunk 7 above)
Comment (alx32) [reply lld/test/MachO/icf-safe-thunks-dwarf.ll:27]: Which one ? There are 2 other `RUN's` that append to the file ( ex: `a_thunks.dSYM >> %t/a_thunks.txt` )

The reason for append + test is to test that the offsets are consistent across the multiple calls (ex: `MERGED_FUN_ADDR`).

File: lld/MachO/SyntheticSections.cpp [hunk 9]
```diff
@@ -1204,6 +1205,17 @@ void SymtabSection::emitEndFunStab(Defined *defined) {
   stabs.emplace_back(std::move(stab));
 }
 
+// Given a pointer to a function symbol, return the symbol that points to the
+// actual function body that will go in the final binary. Generally this is the
+// symbol itself, but if the symbol was folded using a thunk, we retrieve the
+// target function body from the thunk.
+Defined *SymtabSection::getFuncBodySym(Defined *originalSym) {
+  if (originalSym->identicalCodeFoldingKind != Symbol::ICFFoldKind::Thunk)
```
[the remark this answers is not in the document]
Comment (ellishg) [reply lld/MachO/SyntheticSections.cpp]: Oh sorry, I meant that you should flip the condition and the return. So that if it is a thunk, we return the thunk. Otherwise we return the original symbol.

File: lld/test/MachO/icf-safe-thunks-dwarf.ll (hunk 7 above)
Comment (kyulee-com) [reply lld/test/MachO/icf-safe-thunks-dwarf.ll:27]: Yeah. I misread `>>` as `>`.

**Review by kyulee-com — APPROVED** [review]
LGTM
